external-crates/move/crates/move-package/src/compilation/compiled_package.rs

```diff
@@ -756,3 +765,123 @@ pub(crate) fn make_source_and_deps_for_compiler(
     };
     Ok((source_package_paths, deps_package_paths))
 }
+
+/// partitions `deps` by whether we need to compile dependent packages with a
+/// prior toolchain (which we find by looking at Move.lock contents) or
+/// whether we can compile them with the current binary.
+fn partition_deps_by_toolchain(
+    deps: Vec<PackagePaths>,
+) -> Result<(Vec<PackagePaths>, Vec<PackagePaths>)> {
+    let mut deps_for_current_compiler = vec![];
+    let mut deps_for_prior_compiler = vec![];
+    for dep in deps {
+        let a_source_path = dep.paths[0].as_str();
+        let root = SourcePackageLayout::try_find_root(Path::new(a_source_path))?;
+        let lock_file = root.join("Move.lock");
+        if !lock_file.exists() {
+            // Q: Behavior to pick when package has no lock file. Right now we choose the current compiler (we could use the "legacy" one instead).
+            deps_for_current_compiler.push(dep);
+            continue;
+        }
+
+        let mut lock_file = File::open(lock_file)?;
+        let toolchain_version = lock_file::schema::ToolchainVersion::read(&mut lock_file)?;
+        match toolchain_version {
+            None => {
+                // Q: Behavior to pick when package has no [move.toolchain-version] info. Right now we choose the current compiler (we could use the "legacy" one instead).
+                deps_for_current_compiler.push(dep)
+            }
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == env!("CARGO_PKG_VERSION") => {
+                // This dependency requires the same compiler version we're on.
+                deps_for_current_compiler.push(dep)
+            }
+            Some(toolchain_version) => {
+                // Compile and mark that we compiled this dep with a prior compiler.
+                download_and_compile(root, toolchain_version)?;
+                deps_for_prior_compiler.push(dep)
```

Comment:
Should we make this a map? Like collect all deps for version X and compile them all at once. 
Or is it not worth it?

Reply:
Echoing this suggestion: assemble each dep by what version it wants, as `Map<Version, Vec<Dep>>` or similar. While it likely isn't necessary for this logic to work, it may allow future extensions (blacklisting versions, allowing users to filter versions, etc.)

Reply:
Noted! I think it makes sense to keep a `Map` that groups by version, but for V0 of the implementation, in the interest of speed, will probably leave "compile all at once by version" on the table for a second iteration.